src/Simplex/FileTransfer/Client.hs

```diff
@@ -85,26 +93,64 @@ defaultXFTPClientConfig :: XFTPClientConfig
+  let HTTP2Client {sessionId, sessionALPN} = http2Client
+      thParams0 = THandleParams {sessionId, blockSize = xftpBlockSize, thVersion = VersionXFTP 1, thAuth = Nothing, implySessId = False, batch = True}
+  logDebug $ "Client negotiated handshake protocol: " <> tshow sessionALPN
+  thParams <- case fromMaybe "xftp/0" sessionALPN of
+    "xftp/0" -> pure thParams0
+    "xftp/1" -> xftpClientHandshakeV1 g serverVRange keyHash http2Client thParams0
+    _ -> throwError $ PCETransportError (TEHandshake VERSION)
+  let c = XFTPClient {http2Client, thParams, transportSession, config}
   atomically $ writeTVar clientVar $ Just c
   pure c
 
+xftpClientHandshakeV1 :: TVar ChaChaDRG -> VersionRangeXFTP -> C.KeyHash -> HTTP2Client -> THandleParamsXFTP -> ExceptT XFTPClientError IO THandleParamsXFTP
+xftpClientHandshakeV1 g serverVRange keyHash@(C.KeyHash kh) c@HTTP2Client {sessionId, serverKey, serverCerts = _todo} thParams0 = do
+  -- step 1: pass initiative to server to get its handshake
+  let helloReq = H.requestNoBody "POST" "/" []
+  HTTP2Response {respBody = HTTP2Body {bodyHead = shsBody'}} <- liftError' (const $ PCEResponseError HANDSHAKE) $ sendRequestDirect c helloReq Nothing
+  shsBody <- liftEitherWith (const $ PCEResponseError HANDSHAKE) $ C.unPad shsBody'
+  XFTPServerHandshake {xftpVersionRange, sessionId = serverSessId, authPubKey = serverAuth} <- liftEitherWith (const $ PCEResponseError HANDSHAKE) $ smpDecode shsBody
+  -- step 2: validate server handshake
+  unless (sessionId == serverSessId) $ throwError $ PCEResponseError SESSION
+  (params, chs) <- case xftpVersionRange `compatibleVersion` serverVRange of
+    Nothing -> throwError $ PCEResponseError HANDSHAKE
+    Just (Compatible v) -> do
+      sk_ <- liftEitherWith (const $ PCEResponseError HANDSHAKE) $ do
+        let (X.CertificateChain cert, exact) = serverAuth
+        case cert of
+          [_leaf, ca] | XV.Fingerprint kh == XV.getFingerprint ca X.HashSHA256 -> pure ()
+          _ -> throwError "bad certificate"
+        pubKey <- C.verifyX509 serverKey exact
+        C.x509ToPublic (pubKey, []) >>= C.pubKey
+      (pk, k) <- atomically $ C.generateKeyPair g
+      let auth = THandleAuth {peerPubKey = sk_, privKey = k}
+      pure (thParams0 {thAuth = Just auth, thVersion = v}, XFTPClientHandshake {xftpVersion = v, keyHash, authPubKey = pk})
+  -- step 3: send client handshake
+  let chsReq = H.requestBuilder "POST" "/" [] $ padXftp (smpEncode chs)
+  HTTP2Response {respBody = HTTP2Body {bodyHead}} <- liftError' (const $ PCEResponseError HANDSHAKE) $ sendRequestDirect c chsReq Nothing
+  -- step 4: validate server confirmation
+  unless (B.null bodyHead) $ throwError $ PCEResponseError HANDSHAKE
+  -- all done, return negotiated params
+  pure params
+  where
+    padXftp bs = either (error "assert: xftpBlockSize < 65k") byteString $ C.pad bs xftpBlockSize
```

Comment:
why don't you want to keep it as failable? It's in ExceptT anyway

Reply:
request building is pure, didn't want to interleave with never-firing handlers.

this place would light up when searching for `xftpBlockSize`-affected code